### PR TITLE
Add Persistent Subnets

### LINKED ISSUE #5696: Suspicious looking enr["attnets"] bitfields in Topaz
# 🐞 Bug Report

### Description

I am noticing anomalies when inspecting subnet subscriptions stored in enr["attnets"].

topaz-peers-attnets

The column labeled `enr_attnets` represent the bits in the attnets bitfield that == TRUE.  I have yet to see a bit > 4 == TRUE for a peer with a fork-digest of `f071c66c`.  

> Note: the one peer with bit 27 set has a fork-digest of `9925efd6` (Schlesi) and could potentially be a lighthouse node.

We could potentially explain why many of them are in the range of 0 to 4 bc:

1. the `enr["attnets"]` field is including regular committee assignments + persistent committee assignment.  (I requested spec clarification for this: https://github.com/ethereum/eth2.0-specs/issues/1774)
2. 16000/128/32 ~= 4. In other words, There are only 4 to 5 committee indices per slot so only the lower subnets will be used (per @djrtwo).  

That being said, I would expect that random persistent committees would show up in the bitfield and we would see some values in the [5,63] range.  Since I didn't see any...I looked at the code and (might) have an explanation.

Starting at SubscribeCommitteeSubnets(): 

https://github.com/prysmaticlabs/prysm/blob/5636cd3ed87ac35a208d4311ef641bc97fff473e/beacon-chain/rpc/validator/attester.go#L183
-> cache.CommitteeIDs.AddAttesterCommiteeID(req.Slots[i], req.CommitteeIds[i])
   https://github.com/prysmaticlabs/prysm/blob/37cba662f1077cd023c6efdb9e85f12acb49b916/beacon-chain/cache/committee_ids.go#L37

If I call: `AddAttesterCommiteeID(0, 10)`, then I (think) that `AddAttesterCommiteeID(slot,committeeIds)` sets: `0101` instead of `0000 0000 0010`

I might have an explanation for why the tests pass.  In subnet_tests.go:
https://github.com/prysmaticlabs/prysm/blob/f880fb4ee404e9dd6a7d5de06c387c8e8ba865b7/beacon-chain/p2p/subnets_test.go#L130

we have 
```
	cache.CommitteeIDs.AddAttesterCommiteeID(0, 10)
	testService.RefreshENR(0)
	time.Sleep(2 * time.Second)

	exists, err = s.FindPeersWithSubnet(2)
	if err != nil {
		t.Fatal(err)
	}
```

I believe this test passes bc earlier bit 1,2,3 were set here:

https://github.com/prysmaticlabs/prysm/blob/f880fb4ee404e9dd6a7d5de06c387c8e8ba865b7/beacon-chain/p2p/subnets_test.go#L42

and tested here:

https://github.com/prysmaticlabs/prysm/blob/f880fb4ee404e9dd6a7d5de06c387c8e8ba865b7/beacon-chain/p2p/subnets_test.go#L109

### PR DESCRIPTION
**Pull request by nisdas**
Resolves #5696  
- [x] Add network config related to long lived epoch subscriptions.
-  [x] Allow facility in rpc server to account for persistent subnets, with each respective validator's pubkey being the key in the cache. 
-  [x] Add methods to cache for persistent subnets.
-  [x] Fix ENR to only use persistent committees.
-  [x] Add relevant tests.

### CODE REVIEW

File: beacon-chain/cache/committee_ids.go [hunk 1]
```diff
@@ -85,3 +92,43 @@ func (c *committeeIDs) GetAggregatorCommitteeIDs(slot uint64) []uint64 {
...
+	defer c.subnetsLock.RUnlock()
+
+	id, duration, ok := c.persistentSubnets.GetWithExpiration(string(pubkey))
+	if !ok {
+		return []uint64{}, ok, time.Time{}
+	}
+	return id.([]uint64), ok, duration
+}
+
+// GetAllCommittees retrieves all the subscribed committees of all the validators
```
Comment (terencechain) [inline beacon-chain/cache/committee_ids.go:109]:
```suggestion
// GetAllCommittees retrieves all the non-expired subscribed committees of all the validators
```

File: beacon-chain/rpc/validator/assignments.go [hunk 2]
```diff
@@ -74,6 +79,9 @@ func (vs *Server) GetDuties(ctx context.Context, req *ethpb.DutiesRequest) (*eth
 			if ok {
 				nextCommitteeIDs = append(nextCommitteeIDs, ca.CommitteeIndex)
 			}
+			// assign relevant validator to subnet
```
Comment (terencechain) [inline beacon-chain/rpc/validator/assignments.go:82]:
```suggestion
			// Assign relevant validator to subnet.

File: beacon-chain/rpc/validator/assignments_test.go [hunk 3]
```diff
@@ -8,6 +8,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/prysmaticlabs/prysm/beacon-chain/cache"
+
```
Comment (rauljordan) [inline beacon-chain/rpc/validator/assignments_test.go:12]:
```suggestion
```

File: beacon-chain/rpc/validator/assignments.go [hunk 4]
```diff
@@ -90,3 +98,25 @@ func (vs *Server) GetDuties(ctx context.Context, req *ethpb.DutiesRequest) (*eth
 func (vs *Server) StreamDuties(stream ethpb.BeaconNodeValidator_StreamDutiesServer) error {
 	return status.Error(codes.Unimplemented, "unimplemented")
 }
+
```
Comment (terencechain) [inline beacon-chain/rpc/validator/assignments.go:101]: A function comment would be nice

File: beacon-chain/rpc/validator/assignments.go [hunk 5]
```diff
@@ -90,3 +98,25 @@ func (vs *Server) GetDuties(ctx context.Context, req *ethpb.DutiesRequest) (*eth
...
+func assignValidatorToSubnet(pubkey []byte, status ethpb.ValidatorStatus) {
+	if status != ethpb.ValidatorStatus_ACTIVE && status != ethpb.ValidatorStatus_EXITING {
```
Comment (terencechain) [inline beacon-chain/rpc/validator/assignments.go:103]: Why is `status != ethpb.ValidatorStatus_EXITING` needed? 
Shouldn't `status != ethpb.ValidatorStatus_ACTIVE` already imply that?

File: beacon-chain/rpc/validator/assignments_test.go (hunk 3 above)
Comment (terencechain) [inline beacon-chain/rpc/validator/assignments_test.go]:
```suggestion
```

File: beacon-chain/sync/subscriber.go [hunk 6]
```diff
@@ -238,8 +239,14 @@ func (r *Service) subscribeDynamicWithSubnets(
 				if r.chainStarted && r.initialSync.Syncing() {
 					continue
 				}
+
+				// Persistent subscriptions from validators
+				persistentSubs := r.persistentCommitteeIndices()
 				// Update desired topic indices for aggregator
 				wantedSubs := r.aggregatorCommitteeIndices(currentSlot)
+
```
Comment (terencechain) [inline beacon-chain/sync/subscriber.go:247]:
```suggestion
```

File: beacon-chain/rpc/validator/assignments.go (hunk 5 above)
Comment (nisdas) [reply beacon-chain/rpc/validator/assignments.go:103]: `ethpb.ValidatorStatus_EXITING` would still be a valid status for us to register subnets to which is why I use the conditional here.

File: beacon-chain/rpc/validator/assignments.go (hunk 4 above)
Comment (nisdas) [reply beacon-chain/rpc/validator/assignments.go:101]: alright done

File: beacon-chain/rpc/validator/assignments.go (hunk 5 above)
Comment (terencechain) [reply beacon-chain/rpc/validator/assignments.go:105]: Makes sense. Thanks

**Pull request merged**